### PR TITLE
SiStripHitEfficiency workflow in the Prompt Calibration Loop

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This is a follow-up to PR https://github.com/cms-sw/cmssw/pull/37530, and applies the necessary change-set to be able to run the workflow sketched in that PR as part of the Prompt Calibration Loop. 
I also updated various unit tests and relval workflows to be able to test the new addition.

#### PR validation:

I've run the following commands to test the `ALCAPROMPT` production:

```console
cmsDriver.py testReAlCa -s ALCA:PromptCalibProdSiStripHitEfficiency --conditions 123X_dataRun2_v2 --scenario pp --data --era Run2_2018 --datatier ALCARECO --eventcontent ALCARECO --processName=ReAlCa -n 100000 --dasquery='file dataset=/StreamExpress/Run2018D-SiStripCalMinBias-Express-v1/ALCARECO run=325172' --nThreads=4 --no_exec 
```

followed by:

```console
cmsDriver.py stepHarvest -s ALCAHARVEST:SiStripHitEff --conditions 123X_dataRun2_v2 --scenario pp --data --era Run2_2018 --filein file:PromptCalibProdSiStripHitEfficiency.root -n -1 --no_exec                        
```
to test the `ALCAHARVESTING` step.
I have also run successfully:
   * `runTheMatrix.py -l 1001.0, 1001.2` 
   * `scram b runtests`

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport, but a backport to 12_3_X for usage in first beams might be desirable.

cc:
@robervalwalsh @mdelcourt @sroychow

### CODE REVIEW

**Comment by mmusich** [comment]
test parameters:
   * enable_test = threading
   * workflow_threading = 1001,0 1001.2

**Comment by mmusich** [comment]
@cmsbuild, please test
   * though I expect 1001.2 to fail

**Comment by iarspider** [comment]
please test

**Comment by mmusich** [comment]
@tvami, I edited 1h ago...

**Comment by smuzaffar** [comment]
test parameters:
   * enable_test = threading
   * workflow_threading = 1001.0

**Comment by smuzaffar** [comment]
bot is failing to change delete the thumbs down. So I have deleted the original comment and added a new one

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/37712

**Comment by tvami** [comment]
test parameters:
  - workflows = 1001.0, 1001.2
  - enable_test = threading
  - workflow_threading = 1001.0, 1001.2
  - pull_request = https://github.com/cms-sw/cmssw/pull/37712

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
Hi @mmusich 
in here https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c516b/24281/runTheMatrixTHREADING-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5+ALCAHARVD7+ALCAHARVD8/DQM_V0001_R000165121__Express__PCLTest__ALCAPROMPT.root

all the TEC, TID plots are empty, is this due to the limited stats?
eff_layer10 is empty too but the rest of the layers have mean eff to be higher than 1 (1.002), is that coming from how the efficiency is defined?

**Comment by mmusich** [comment]
@tvami 

> all the TEC, TID plots are empty, is this due to the limited stats?

I don't see empty plots:

Screenshot from 2022-04-28 08-03-54

> the rest of the layers have mean eff to be higher than 1 (1.002), is that coming from how the efficiency is defined?

if you are referring to the mean value, that's an artifact of the binning. With such low stats everything is in the efficiency = 1 bin:

Screenshot from 2022-04-28 08-04-26

**Comment by tvami** [comment]
Screen Shot 2022-04-28 at 09 14 09

This is the plot I meant.
Thanks for the explanation for the rest

**Comment by kskovpen** [comment]
+pdmv

**Comment by tvami** [comment]
Hi @cms-sw/upgrade-l2 this PR changes some workflows by adding a new producer, would you please have a look and sign? Thanks!

**Comment by AdrianoDee** [comment]
+Upgrade

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import FWCore.ParameterSet.Config as cms
+
+# ------------------------------------------------------------------------------
+# configure a filter to run only on the events selected by TkAlMinBias AlcaReco
+import copy
+from HLTrigger.HLTfilters.hltHighLevel_cfi import *
+ALCARECOCalMinBiasFilterForSiStripHitEff = copy.deepcopy(hltHighLevel)
+ALCARECOCalMinBiasFilterForSiStripHitEff.HLTPaths = ['pathALCARECOSiStripCalMinBias']
+ALCARECOCalMinBiasFilterForSiStripHitEff.throw = True ## dont throw on unknown path names
+ALCARECOCalMinBiasFilterForSiStripHitEff.TriggerResultsTag = cms.InputTag("TriggerResults","","RECO")
```
Comment (perrotta) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py:10]:
```suggestion
ALCARECOCalMinBiasFilterForSiStripHitEff = hltHighLevel.clone(
        HLTPaths = ['pathALCARECOSiStripCalMinBias'],
        throw = True, ## throw on unknown path names
        TriggerResultsTag = ("TriggerResults","","RECO")
)
```
(I also changed the comment to the `thow` parameter...)

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+from RecoVertex.BeamSpotProducer.BeamSpot_cff import *
+from RecoTracker.IterativeTracking.InitialStep_cff import *
+from RecoTracker.Configuration.RecoTrackerP5_cff import *
+from RecoTracker.TrackProducer.TrackRefitter_cfi import *
+from DQM.SiStripCommon.TkHistoMap_cff import *
+from RecoTracker.MeasurementDet.MeasurementTrackerEventProducer_cfi import *
+
+ALCARECOMonitoringTracksRefit = TrackRefitter.clone(src = cms.InputTag("ALCARECOMonitoringTracks"),
+                                                     NavigationSchool = cms.string("")
+                                                     )
```
Comment (perrotta) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py:35]:
```suggestion
ALCARECOMonitoringTracksRefit = TrackRefitter.clone(
         src = "ALCARECOMonitoringTracks",
         NavigationSchool = cms.string("")
)
```

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ALCARECOTrackFilterRefit = cms.Sequence(ALCARECOMonitoringTracks +
+                                        MeasurementTrackerEvent +
+                                        offlineBeamSpot +
+                                        ALCARECOMonitoringTracksRefit)
+
+# ------------------------------------------------------------------------------
+# This is the module actually doing the calibration
+from CalibTracker.SiStripHitEfficiency.siStripHitEfficiencyWorker_cfi import siStripHitEfficiencyWorker
+ALCARECOSiStripHitEff =  siStripHitEfficiencyWorker.clone(
+    lumiScalers=cms.InputTag("scalersRawToDigi"),
```
Comment (perrotta) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py:49]: Please remove type specification for the cloned parameters, which is a safer way to overwrite them

**Comment by mmusich** [comment]
@perrotta,as I mentioned elsewhere I can't update this branch earlier than Wednesday. If it is so, will this PR miss the next pre-release?
I am asking because while your comments are valid and will be addressed in due time, they are not touching core functionalities of the worklows that we would need to be tested at tier0 before backporting in 12.3.x (it will be needed for first collisions).
Thanks for the clarification.

**Comment by qliphy** [comment]
@mmusich  The next 12_4_X pre-release will be built in ~ 10 days from now.  So please try to address above comments later this week at your earlier convenience.

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py (hunk 1 above)
Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py]: done at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_Output_cff.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
+
+OutALCARECOPromptCalibProdSiStripHitEfficiency_noDrop = cms.PSet(
+    SelectEvents = cms.untracked.PSet(
+        SelectEvents = cms.vstring('pathALCARECOPromptCalibProdSiStripHitEfficiency')),
+    outputCommands = cms.untracked.vstring('keep *_MEtoEDMConvertSiStripHitEff_*_*'))
+
+import copy
+OutALCARECOPromptCalibProdSiStripHitEfficiency=copy.deepcopy(OutALCARECOPromptCalibProdSiStripHitEfficiency_noDrop)
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_Output_cff.py]: done at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

File: CalibTracker/SiStripHitEfficiency/plugins/SiStripHitEfficiencyWorker.cc [hunk 5]
```diff
@@ -904,7 +908,35 @@ void SiStripHitEfficiencyWorker::endJob() {
...
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("lumiScalers", edm::InputTag{"scalersRawToDigi"});
+  desc.add<edm::InputTag>("commonMode", edm::InputTag{"siStripDigis", "CommonMode"});
+  desc.add<edm::InputTag>("combinatorialTracks", edm::InputTag{"generalTracks"});
+  desc.add<edm::InputTag>("trajectories", edm::InputTag{"generalTracks"});
+  desc.add<edm::InputTag>("siStripClusters", edm::InputTag{"siStripClusters"});
+  desc.add<edm::InputTag>("siStripDigis", edm::InputTag{"siStripDigis"});
+  desc.add<edm::InputTag>("trackerEvent", edm::InputTag{"MeasurementTrackerEvent"});
+  desc.add<int>("Layer", 0);
+  desc.add<bool>("Debug", false);
```
[the remark this answers is not in the document]
Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/SiStripHitEfficiencyWorker.cc]: cleaned up at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py (hunk 2 above)
Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py]: done at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py (hunk 3 above)
Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripHitEfficiency_cff.py]: done at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

File: Calibration/TkAlCaRecoProducers/python/AlcaSiStripHitEfficiencyHarvester_cfi.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import FWCore.ParameterSet.Config as cms
+
+from DQM.SiStripCommon.TkHistoMap_cff import *
+from CalibTracker.SiStripHitEfficiency.siStripHitEfficiencyHarvester_cfi import siStripHitEfficiencyHarvester
+alcasiStripHitEfficiencyHarvester =  siStripHitEfficiencyHarvester.clone(
+    isAtPCL             = cms.bool(True),
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/AlcaSiStripHitEfficiencyHarvester_cfi.py]: done at https://github.com/cms-sw/cmssw/pull/37708/commits/5e79159a730dd71c6faf73ff474272440903dace

**Comment by mmusich** [comment]
urgent:
   * just to flag it's necessary for the next pre-release.

**Comment by tvami** [comment]
@cms-sw/upgrade-l2 @cms-sw/pdmv-l2 would you please re-sign this PR? thanks!

**Comment by srimanob** [comment]
+Upgrade

resign.

**Comment by tvami** [comment]
@cms-sw/orp-l2 this is essentially fully signed now!

**Comment by perrotta** [comment]
+1
- Thank you @mmusich

**Pull request merged**